### PR TITLE
Bug fix and added some test

### PR DESCRIPTION
**Pull request by 0xshora**
Update README

Bug fix
- SettleGameSystem.sol
- ChaosBurnSys.sol

Add some test code

### CODE REVIEW

File: packages/contracts/src/systems/JoinSystem.sol [hunk 1]
```diff
@@ -24,36 +26,42 @@ contract JoinSystem is System {
...
         if (playerData.status != PlayerStatus.UNINITIATED) {
             return false;
         }
+        
+        // check if the time is correct
+        if (gameData.startTime == 0) {
+            gameData.startTime = uint32(block.timestamp) + Config.getGameStartWaitingTime();
+            Game.set(gameData);
+        } else if (block.timestamp >= gameData.startTime) {
+            return false;
```
Comment (noyyyy) [inline packages/contracts/src/systems/JoinSystem.sol]: I think it's better to revert, to tell the user he failed for this transaction as we hardly see the return value.

File: packages/contracts/src/systems/SettleGameSystem.sol [hunk 2]
```diff
@@ -37,54 +38,99 @@ contract SettleGameSystem is System {
...
-                Game.deleteRecord();
             }
+            Game.deleteRecord();
             return true;
         }
 
-        // if there is no player,
-        // distribute the prize to the players who were alive in last round
-        if (numAlives == 0) {
+        if (numAlives != 0) {
```
Comment (noyyyy) [inline packages/contracts/src/systems/SettleGameSystem.sol:59]: i suggest revert

File: packages/contracts/src/systems/SettleGameSystem.sol [hunk 3]
```diff
@@ -37,54 +38,99 @@ contract SettleGameSystem is System {
...
+
+        // if there is no player,
+        // distribute the prize to the players who were alive in last round
+        for (uint32 i = 0; i < all_players.length; i++) {
+            address playerAddr = all_players[i];
+
+            if (Player.getLastCheckedTime(playerAddr) > penultimate_deadline
+                    && Player.getLastCheckedTime(playerAddr) < last_deadline) {
+                if (numLastAlives != 0) {
+                    (bool success, ) = payable(playerAddr).call{value: prize / numLastAlives}("");
```
Comment (noyyyy) [inline packages/contracts/src/systems/SettleGameSystem.sol:121]: we should check whether the success is true. Or we can directly use openzeplin's `Address.sendValue`

File: packages/contracts/src/systems/SettleGameSystem.sol [hunk 4]
```diff
@@ -37,54 +38,99 @@ contract SettleGameSystem is System {
...
+                    require(success, "SettleGameSystem: failed to transfer prize to player");
+                } else {
+                    (bool success, ) = payable(playerAddr).call{value: prize / all_players.length}("");
+                    require(success, "SettleGameSystem: failed to transfer prize to player");
                 }
-                // delete player record
                 Player.deleteRecord(playerAddr);
-                Game.deleteRecord();
```
Comment (noyyyy) [inline packages/contracts/src/systems/SettleGameSystem.sol:83]: shouldn't remove this line, or play cannot join for new game

File: packages/contracts/src/template/ChaosBurnSys.sol [hunk 5]
```diff
@@ -1,32 +1,59 @@
...
+import {Game, GameData} from "../codegen/index.sol";
 import {PlayerStatus} from "../codegen/common.sol";
+import {IWorld} from "src/codegen/world/IWorld.sol";
+import {SystemSwitch} from "@latticexyz/world-modules/src/utils/SystemSwitch.sol";
 
 /**
  * @notice this system make the amount of token into a random things.
  */
 contract ChaosBurnSys is System {
+    mapping(uint256 => bool) public hasBurnedThisRound;
```
Comment (noyyyy) [inline packages/contracts/src/template/ChaosBurnSys.sol]: it doesn't work at all. We already have defined a value in `mud.config.ts`

Comment (noyyyy) [reply packages/contracts/src/template/ChaosBurnSys.sol]: I think it's best not to write like this. In mud, storage defination and logic is seperate, we'd better obey this rule

File: packages/contracts/src/systems/SettleGameSystem.sol (hunk 3 above)
Comment (0xshora) [reply packages/contracts/src/systems/SettleGameSystem.sol:121]: https://github.com/StayWarmLabs/StayWarm/blob/dev/syora/packages/contracts/src/systems/SettleGameSystem.sol#L122

Does this line work for checking?

File: packages/contracts/src/systems/SettleGameSystem.sol (hunk 4 above)
Comment (0xshora) [reply packages/contracts/src/systems/SettleGameSystem.sol:83]: We have `Game.deleteRecord()` at the end of settlements.
Like:
https://github.com/StayWarmLabs/StayWarm/blob/dev/syora/packages/contracts/src/systems/SettleGameSystem.sol#L132

File: packages/contracts/src/template/ChaosBurnSys.sol (hunk 5 above)
Comment (0xshora) [reply packages/contracts/src/template/ChaosBurnSys.sol]: Hmmmm... I see.
Then, do you have any idea to implement this?: I just want to limit players to change burn_amount once per round.

Or, I wonder how can I add new components(storage) for the new system.

Comment (noyyyy) [reply packages/contracts/src/template/ChaosBurnSys.sol]: > Hmmmm... I see. Then, do you have any idea to implement this?: I just want to limit players to change burn_amount once per round.
> 
> Or, I wonder how can I add new components(storage) for the new system.

To add new storage value, we can directly modify `mud.config.ts`, then gererate new library, then import new library to system and use it. The storage in mud is abstract. You can take a look at the generated library to see how it impl storage.

Comment (noyyyy) [reply packages/contracts/src/template/ChaosBurnSys.sol]: > Hmmmm... I see. Then, do you have any idea to implement this?: I just want to limit players to change burn_amount once per round.
> 
> Or, I wonder how can I add new components(storage) for the new system.

If we implment this demand, we need a dedicated `mud.config.ts`. I prefer not to implement it in this repo.

Comment (noyyyy) [reply packages/contracts/src/template/ChaosBurnSys.sol]: If a player want to add new storage, he can fork the repo and modify `mud.config.ts`. Once his proposal is passed, the new `mud.config.ts` can be merged to the the game session repo.

**Review by noyyyy — APPROVED** [review]
We can merge it first. And I can handle the deploy things if you're still stuck.

**Pull request merged**